pageserver/src/tenant/timeline.rs

```diff
@@ -2861,7 +2861,14 @@ impl Timeline {
             self.metrics
                 .resident_physical_size_gauge
                 .add(metadata.len());
-            updates.insert_historic(Arc::new(l));
+            let l = Arc::new(l);
+            if updates.contains(l.clone()) {
+                error!(
+                    "Attempt to create duplicated image layer {}",
+                    l.filename().file_name()
+                );
+            }
+            updates.insert_historic(l);
```

Comment:
I think this is a bad pattern to assert in the deep and *ask* before adding. Bad because we should now remember to do this every time.

a good counterpart for the LayerMap to follow in semantics would be the HashSet or BTreeSet. then the `insert` should return bool (true = changed, false = not modified as in it was already included). `HashSet::insert`: https://doc.rust-lang.org/stable/std/collections/hash_set/struct.HashSet.html#method.insert

Instead of:
- asserting in the deep => return error up (to here basically)
- assert at the callsite => `assert!(updates.insert_historic(l));`

Reply:
This will probably infect all of the callsites. I wonder if we should add more error handling in the case that we do in fact add a duplicate in?

Asserting is not yet really supported because we lose the ability to probably even ignore the tenant.

Reply:
Replaced

Reply:
Insert is used by replace, so I can not skip insertion f value already exists.
Also LayerMap is actually map, not set. 
So `insert` method should return previous value.
But we do not nee it, for our purposes it is more convenient to return bool.